README.md
# Purpose
Improve efficiency and accuracy for data between clinical outcomes databases and source databases.

# Getting Started
**Files needed:**
1. source_data.csv (Source database sheet save as *UTF-8* encoding csv file)
2. data_augmented_xl.csv (one of the output files from outcomes database main augmentation)

**Scripts needed:**
1. augment.source 02162023_v5.R
2. source.functions.R

**Directories:**<br>
Save the files in ***csv*** folder of the database you are currently working on, and the scripts in ***r*** folder.<br>
<img width="600" alt="screenshot" src="https://user-images.githubusercontent.com/49822948/223860085-6c833ba4-e071-45b0-a2c1-09055c93e10b.PNG">

### Usage
Make sure source_data.csv is in *UTF-8* encoding in order to avoid special characters. A few things listed below worth noticing while augmenting the source database.
 
Lines 95-98 are coded to remove internal columns. Remove anything after **database** column.
```r
# Remove internal use columns (not in specs), please ignore date.transform column
# The 2 default columns are: entry, qc
names(datSource)
internalCols <- c("entry","QC") # adjust internal colnames as needed
```

Lines 130-135 are checking named ranges on search, include, reason, e.copy, and database columns. Please harmonize and change it as it see fits.
```r
# Run line-by-line
unique(datSource$search[!datSource$search %in% searchType])
unique(datSource$include[!datSource$include %in% yesNoUnclear])
unique(datSource$reason[!datSource$reason %in% reasonType])
unique(datSource$e.copy[!datSource$e.copy %in% yesNo])
unique(datSource$database[!datSource$database %in% yesNo])
```

Lines 210-228 are making sure the treatment info are not missing. Usually, the x dimension of datUniRef and datUniTrt should be the same. But when a reference/study has only subarms, strata, or control arms, the x dimension could vary. Please check the outcomes database to verify first. And uncomment the codes in line 218-228 to add the treatment information back.
```r
# datUniRef and datUniTrt datasets should have the same x dimension \\\datUniCtrl could be missing due to single arm studies
dim(datUniRef)
dim(datUniCtrl)
dim(datUniTrt)
# can use the following code to find out which is missing in datUniTrt and look into augemented db
datUniRef$source.number[!datUniRef$source.number %in% datUniTrt$source.number]
# !!!Use the next chunk of codes if we have missing in datUniTrt!!!
 
# # Join treatment if the arms are subarms, strata only reference (have to check first in outcomes DB)
# tmp <- datUniRef$source.number[datUniRef$source.number %in% datUniTrt$source.number == FALSE]
# tmp.trt <- OutcomesDB %>%
#   filter(source.number %in% tmp) %>%
#   filter(!duplicated(treatment)) %>%
#   group_by(source.number) %>%
#   mutate(treatment.join = paste(treatment, collapse = ", ")) %>%
#   filter(!duplicated(source.number)) %>%
#   select(source.number, treatment.join)
# 
# datUniTrt <- rbind(datUniTrt, tmp.trt)
```

Line 299 is to check if information for references more than 1 study has been properly concatenated.
```r
view(check.multi.study)
```

Please enjoy!


CHANGELOG.md
[v.2.0.0] - 2023-03-08


## Added
- Web Scraping
- Extract, transform, and load it to our existing database
## Changed

## Fixed

source.functions.R
# DESCRIPTION -------------------------------------------------------------
# Purpose        : Source Augmentation Functions Script
# Script Version : 17 February 2021
# Database       :
# Created by     : Cynthia Yeh
#-------------------------------------------------------------------------
# Last run by  : (name) - Certara Confidential
# Date run     : 
# -------------------------------------------------------------------------
# Software : Version 1.1.383 - 2009-2017 RStudio, Inc.
#            R version 4.0.4
# Platform : Lenovo ThinkPad T490
# Environment : Windows 10 Enterprise, Intel(R) Core(TM) i7-10510U CPU @ 1.80 GHz 64-bit OS



# Generic Checks

checkReqSourceCols = function(data,show.details=T){
  req.cols = c('search',
               'search.date',
               'url',
               'source.number',
               'authors',
               'publication.year',
               'title',
               'journal',
               'volume',
               'pages',
               'abstract',
               'include',
               'reason',
               'reason.descr',
               'study',
               'registry.study.id',
               'treatment',
               'control',
               'n.study',
               'n.arm',
               'e.copy',
               'database'
  )
  miss.cols = NULL
  miss.cols = req.cols[!(req.cols %in% colnames(data))]
  k = length(miss.cols)
  if (k == 0) {
    message("All required columns are present.")
  } else {
    if (show.details == T) {
      message(paste("The following",k,"columns are missing and required:"))
      miss.cols
    } else {
      message(paste("There are",k,"columns missing."))
    }
  }
}

nonnum.unique = function(x){
  nonnum = is.na(suppressWarnings(as.numeric(as.character(x))))
  n = which(nonnum & !is.na(x))
  n.lab = paste("(",n,")",sep = '')
  if (length(n) == 0) {
    "all numeric"
  } else {
    paste(as.character(x[n]),n.lab,sep = '')
  }
}

checkNumSourceCols = function(data,show.details=T){
  num.cols = c('source.number', # added by cy
               'publication.year')
  num.cols2 = num.cols[num.cols %in% colnames(data)]
  df = data[,num.cols2]
  t2 = sapply(df,nonnum.unique) 
  t3 = t2[t2 != "all numeric"]
  k = length(t3)
  if (k == 0) {
    message("There are no columns with unallowed non-numerics.")
  } else {
    if (show.details == T) {
      message(paste("Warning: The following",k,"variables have unallowed non-numeric entries. Please correct them:"))
      names(t3)
    } else {
      message(paste("There are",k,"variables that have unallowed missings."))
    } 
  }
}

# Named Ranges Checks

searchType <- c("pubmed","clinicaltrials.gov","fda","ema","conference","clinical study report",
                "company website","cross-reference","web search")
yesNoUnclear<-c("yes","no","unclear")
yesNo<-c("yes","no")
reasonType<-c("","indication","design","treatment","comparison","endpoints","population",
              "sample size","availability","reliability","secondary reference","other", "study duration")

# Check Database

checkSourceYes = function(data){
  if (length(data > 0)) {
    message("The following references are marked as database = yes in source DB, but not found in outcomes DB. 
            Please double check and correct them:")
    sort(data)
  } else {
    message("Perfect! Let's go.")
  }
}

checkSourceNo = function(data){
  if (nrow(data) > 0) {
    message("These references are pending or not curated yet")
    sort(data$source.number)
  } else {
    message("Great!!! All included references are curated.")
  }
}

# Dataset Trimming

TrimData <- function(dataset, ...){
  col_vars <- quos(...)
  datUni <- dataset  %>%
    group_by(!!!col_vars) %>%
    mutate(unique.id = cur_group_id()) %>%
    filter(!duplicated(unique.id))
}

# Identify Multiple Studies Sharing Single Source Number

checkMultiStudy = function(data){
  if (sum(duplicated(data$source.number)) > 0) {
    message("The following source numbers have more than 1 study")
    sort(unique(data$source.number[duplicated(data$source.number)]))
    } else {
      message("you are good to go! :D")
    }
}

# Replace Information From Augmented Database

ReplaceInfo = function (x) {
  datSource[[x]] <- ifelse(!is.na(datSource[[paste0(x, ".join")]]), datSource[[paste0(x, ".join")]], datSource[[x]])
  return(datSource[[x]])
}


# Warning Messages

ManualFill = function(data){
  if(length(data) > 0) {
    message("Please manually fill back info for following ", paste0(length(miss.in.source)),
          " references: ",
          paste(data, collapse = ", "))
  }
}

# Find the position for missing URLs, if there is any

MissingURLs = function(data){
  if (sum(is.na(data$url[data$database == "yes"]) > 0)) {
    message("Identify some missing URLs, please check them in the source DB")
    which(is.na(data$url[data$database == "yes"]))
    } else {
      message("No missing URLs. You are good to go!")
    }
}

# Check unallowed missings

MissingInfo = function(data){
  dbCols = c("search", "search.date", "url", "include", "e.copy")
  df = data[,dbCols]
  t2 = sapply(df, function(x) sum(is.na(x))) 
  t3 = sum(t2)
  if (t3 == 0) {
    message("No unallowed missings. You are good to go!")
    } else {
    message(paste("Warning: The following variables have unallowed missings. Please correct them:"))
    t2[t2 > 0]
    }
  }


src/augment.source 03032021_v3.R
# DESCRIPTION -------------------------------------------------------------
# Purpose        : Source Augmentation Script
# Script Version : 17 February 2021
# Database       :
# Created by     : Cynthia Yeh
#-------------------------------------------------------------------------
# Last run by  : (name) - Certara Confidential
# Date run     : 07 April 2020
# -------------------------------------------------------------------------
# Software : Version 1.1.383 - 2009-2017 RStudio, Inc.
#            R version 4.0.4
# Platform : Lenovo ThinkPad T490
# Environment : Windows 10 Enterprise, Intel(R) Core(TM) i7-10510U CPU @ 1.80 GHz 64-bit OS



# DATASET NAMING -------------------------------------------------------------
# OutcomesDB      <- augmented working DB
# SourceDB        <- source DB
# datSource       <- transform from SourceDB and will be the final augmented dataset
# datUniRef       <- dataset that contains unique references
# datUniCtrl      <-dataset that contains unique control drugs for studies
# datUniTrt       <- dataset that contains unique interventional drugs for studies
# datUniRef.join  <- Merged dataset from datUniRef/datUniCtrl/datUniTrt by source.number


# DEFAULTS ----------------------------------------------------------

rm(list = ls())

# Dependencies
#install.packages("dplyr")
#install.packages("purrr")
#install.packages("tidyverse")
#install.packages("lubridate")
library(dplyr)
library(purrr)
library(tidyverse)
library(lubridate)

# Directories
dirHome <- dirname(rstudioapi::getSourceEditorContext()$path)
setwd(dirHome)

# Check that working directory is correct
getwd()
dirSource <- '../csv'
dirResults <- '../csv'

# Files
sourcedb <- 'source_data.csv'
augmentdb<-'data_augmented_xl.csv'


# DATA IMPORT ----------------------------------------------------------------

OutcomesDB <- read.csv(file = paste(dirSource, augmentdb, sep = "/"), header = TRUE, fill = TRUE,
                       as.is = TRUE, sep = ",", na.strings = c(".","N/A"))
SourceDB <- read.csv(file = paste(dirSource, sourcedb, sep = "/"), header = TRUE, fill = TRUE,
                     as.is = TRUE, sep = ",", na.strings = c(".","N/A"), strip.white = TRUE)
# exclude empty rows
OutcomesDB <- filter(OutcomesDB, !source.number == "")
SourceDB <- filter(SourceDB, !source.number == "")

source('source.functions.R')

# Check original dimensions of database
dim(OutcomesDB)
dim(SourceDB)

# DATA PREP ---------------------------------------------------------------------

datSource <- SourceDB
dim(datSource) 

### Date transformation (datSource)
# Transform date format to yyyy-mm-dd
datSource$date.transform <- mdy(datSource$search.date)

# Convert date columns to desired format
dates <- c("search.date", "date.transform")
for (k in dates) {
  datSource[,k] <- paste('="', as.character(datSource[,k]), '"', sep = '') 
}

datSource$search.date <- ifelse(datSource$date.transform == "=\"NA\"", 
                                as.character(datSource$search.date), 
                                datSource$date.transform)

### Data Cleaning (datSource)
# Remove internal use columns (not in specs), please ignore date.transform column
# The 2 default columns are: entry, qc
names(datSource)
internalCols <- c("entry", "qc") # adjust internal colnames as needed

datSource <- datSource %>%
  mutate(across(where(is.character), trimws),
         across(c(search, include, reason, e.copy, database), tolower)) %>%
  select(!all_of(internalCols) & !date.transform)

names(datSource) # database should be the last column

### Subset needed comlumns (OutcomesDB)
OutcomesDB <- OutcomesDB %>%
  select(url:publication.type,
         study,
         registry.study.id,
         n.study,
         arm,
         control.arm,
         randomized.drug)


# DATA FORMAT CHECKS ----------------------------------------------------------------

# Check required columns 
checkReqSourceCols(datSource)
checkNumSourceCols(datSource)

# Use following for unallowed non-numeric entries. Please uncomment following 2 lines and change the colnames in x.
#x <- c("source.number", "publication.year")
#datSource[x] <- sapply(datSource[x], as.numeric)


# NAMED RANGES ------------------------------------------------------------------

# Run line-by-line
unique(datSource$search[!datSource$search %in% searchType])
unique(datSource$include[!datSource$include %in% yesNoUnclear])
unique(datSource$reason[!datSource$reason %in% reasonType])
unique(datSource$e.copy[!datSource$e.copy %in% yesNo])
unique(datSource$database[!datSource$database %in% yesNo])


# URL HARMONIZATION ---------------------------------------------------------------------- (need improvements)

# Generate PubMed/clintrials URLs
datSource$url[datSource$search == "pubmed"] <- paste0("https://www.ncbi.nlm.nih.gov/pubmed/",
                                                      datSource$source.number[datSource$search == "pubmed"])
datSource$url[datSource$search == "clinicaltrials.gov"] <- paste0("https://clinicaltrials.gov/ct2/show/",
                                                                  datSource$registry.study.id[datSource$search == "clinicaltrials.gov"])
# Remove everything that comes after a comma in the URL
datSource$url <- sapply(strsplit(datSource$url, ","), "[", 1)


# REFERENCES CHECKS -----------------------------------------------------------------

wDB.sNumber <- unique(OutcomesDB$source.number)
sDB.database.yes <- unique(datSource$source.number[datSource$database == "yes"])
sDB.database.no <- unique(datSource$source.number[datSource$database != "yes"])

# Check database = yes in source DB
wDB.no <- sDB.database.yes[!sDB.database.yes %in% wDB.sNumber]
checkSourceYes(wDB.no)

# Check database == no in source DB
wDB.yes <- sDB.database.no[sDB.database.no %in% wDB.sNumber]
datSource$database[datSource$source.number %in% wDB.yes] <- "yes"

# Check the rest of database == no (& include == yes) in Source DB
sDB.include.yes <- datSource %>%
  filter(include == "yes", database != "yes") %>%
  select(source.number)
checkSourceNo(sDB.include.yes)


# CREATE DATASETS ----------------------------------------------------------------

### Create datasets for unique references, unique control drugs, and unique randomized drugs for studies ###

### Dataset - unique referenes that contains maximum arm for each study
datUniRef <- TrimData(OutcomesDB, source.number, study, arm)
datUniRef$arm <- gsub(".*\\+|[^0-9]", "", datUniRef$arm)
datUniRef <- datUniRef %>%
  group_by(source.number, study) %>%
  slice_max(arm)

# Identify duplicated source.number for more than 1 study, if any
checkMultiStudy(datUniRef)
refs.multiple.study <- unique(datUniRef$source.number[duplicated(datUniRef$source.number)])

# Combine multiple studies information
datUniRef <- datUniRef %>%
  group_by(source.number) %>%
  mutate(across(c(study, registry.study.id, n.study, arm),
                function(x) paste(unique(x[!is.na(x) & x != ""]), collapse = " and "))) %>%
  filter(!duplicated(source.number))

### Dataset - unique control drugs dataset
datUniCtrl <- filter(OutcomesDB, !grepl("[A-z]", arm),            # exclude subarms
                     control.arm == "yes")                   # include control arms only
datUniCtrl <- TrimData(datUniCtrl, source.number, study, randomized.drug)

datUniCtrl <- datUniCtrl %>%
  group_by(source.number) %>%
  filter(!duplicated(randomized.drug)) %>%
  mutate(control.join = paste(randomized.drug, collapse = ", ")) %>%
  filter(!duplicated(source.number)) %>%
  select(source.number, control.join)


### Dataset - unique interventional drugs dataset
datUniTrt <- filter(OutcomesDB, !grepl("[A-z]", arm),             # exclude subarms
                    control.arm == "no")                     # include experimental arms only
datUniTrt <- TrimData(datUniTrt, source.number, study, randomized.drug)

datUniTrt <- datUniTrt %>%
  group_by(source.number) %>%
  filter(!duplicated(randomized.drug)) %>%
  mutate(treatment.join = paste(randomized.drug, collapse = ", ")) %>%
  filter(!duplicated(source.number)) %>%
  select(source.number, treatment.join)

# datUniRef and datUniTrt datasets should have the same x dimension \\\datUniCtrl could be missing due to single arm studies
dim(datUniRef)
dim(datUniCtrl)
dim(datUniTrt)


# FILL IN MISSING INFORMATION ---------------------------------------------------------------------------

# Merge datUniRef/datUniCtrl/datUniTrt
cols.join <- c("url",
               "authors",
               "publication.year",
               "title",
               "journal",
               "volume",
               "pages",
               "study",
               "registry.study.id",
               "n.study",
               "arm")

datUniRef.join <- datUniRef %>%
  select(source.number, all_of(cols.join)) %>%
  rename(n.arm = arm) %>%
  rename_at(vars(-source.number), paste0, ".join") %>%
  list(datUniCtrl, datUniTrt) %>%
  reduce(left_join, by = "source.number")

# For oncology database single arm study (can also run it in general database)
datUniRef.join$control.join[is.na(datUniRef.join$control.join)] <- "no control"


### Check if it's in the augmented database, but not in the source database ###
miss.in.source <- wDB.sNumber[!wDB.sNumber %in% datSource$source.number]

# Add missing refs' source numbers to datUniRef.join for filling information later  
for (miss.info in miss.in.source) {
  datSource <- datSource %>%
    add_row(source.number = miss.info, database = "yes")
}


### All augmentated DB source.number should be in the source database now, proceed to fill in information ###
cols.join <- gsub(".join", "", names(datUniRef.join[, names(datUniRef.join) != "source.number"]))
cols.join

datSource <- datSource %>%
  left_join(datUniRef.join, by = "source.number")

# Fill in info from augmented outcome DB
for(names in cols.join) {
  datSource[[names]] <- ReplaceInfo(names)
}

# CHECK INFORMATION FOR REFERENCES MORE THAN 1 STUDY -------------------------------------------------
check.multi.study <- datSource %>%
  filter(source.number %in% refs.multiple.study) %>%
  select(source.number,
         study, 
         registry.study.id,
         treatment,
         control,
         n.study,
         n.arm)
view(check.multi.study)

# DATA FINAL CLEANING --------------------------------------------------------------------------------

#Unify study name format (remove quotation marks)
datSource$study <- gsub('=|"', "", datSource$study)

# Final cleaning and save to file
datSource <- select(datSource, !ends_with("join"))
write.csv(datSource, paste(dirResults,'source_augmented.csv',sep = '/'), na = "", row.names = FALSE)


# MESSAGES ---------------------------------------------------------------

# Find the position for any new information needed manually
#miss.in.source <- wDB.sNumber[!wDB.sNumber %in% datSource$source.number]
#wDB.sNumber[!wDB.sNumber %in% datSource$source.number]
ManualFill(miss.in.source)

# Find the position for missing URLs, if there is any
MissingURLs(datSource)


# END OF CODE -----------------------------------------------------------------------------

unique(datSource$database[!datSource$database %in% yesNo])


augment.source 02162023_v5.R
# DESCRIPTION -------------------------------------------------------------
# Purpose        : Source Augmentation Script
# Script Version : 2 December 2022
# Database       :
# Created by     : Cynthia Yeh
#-------------------------------------------------------------------------
# Last run by  : (name) - Certara Confidential
# Date run     : 07 April 2020
# -------------------------------------------------------------------------
# Software : Version 1.1.383 - 2009-2017 RStudio, Inc.
#            R version 4.0.4
# Platform : Lenovo ThinkPad T490
# Environment : Windows 10 Enterprise, Intel(R) Core(TM) i7-10510U CPU @ 1.80 GHz 64-bit OS



# DATASET NAMING -------------------------------------------------------------
# OutcomesDB      <- augmented working DB
# SourceDB        <- source DB
# datSource       <- transform from SourceDB and will be the final augmented dataset
# datUniRef       <- dataset that contains unique references
# datUniCtrl      <- dataset that contains unique control drugs for studies
# datUniTrt       <- dataset that contains unique interventional drugs for studies
# datUniRef.join  <- Merged dataset from datUniRef/datUniCtrl/datUniTrt by source.number

# NOTES FOR CHANGE ----------------------------------------------------------
# "treatment" <- "randomized.drug"

# DEFAULTS ----------------------------------------------------------

rm(list = ls())

# Dependencies
#install.packages("dplyr")
#install.packages("purrr")
#install.packages("tidyverse")
#install.packages("lubridate")
library(dplyr)
library(purrr)
library(tidyverse)
library(lubridate)

# Directories
dirHome <- dirname(rstudioapi::getSourceEditorContext()$path)
setwd(dirHome)

# Check that working directory is correct
getwd()
dirSource <- '../csv'
dirResults <- '../csv'

# Files
sourcedb <- 'source_data.csv'
augmentdb<-'data_augmented_xl.csv'


# DATA IMPORT ----------------------------------------------------------------

OutcomesDB <- read.csv(file = paste(dirSource, augmentdb, sep = "/"), header = TRUE, fill = TRUE,
                       as.is = TRUE, sep = ",", na.strings = c(".","N/A"))
SourceDB <- read.csv(file = paste(dirSource, sourcedb, sep = "/"), header = TRUE, fill = TRUE,
                     as.is = TRUE, sep = ",", na.strings = c(".","N/A"), strip.white = TRUE, encoding = "UTF-8")
# exclude empty rows
OutcomesDB <- filter(OutcomesDB, !source.number == "")
SourceDB <- filter(SourceDB, !source.number == "")

source('source.functions.R')

# Check original dimensions of database
dim(OutcomesDB)
dim(SourceDB)

# DATA PREP ---------------------------------------------------------------------

datSource <- SourceDB
dim(datSource) 

### Date transformation (datSource)
# Transform date format to yyyy-mm-dd
datSource$date.transform <- mdy(datSource$search.date)

# Convert date columns to desired format
dates <- c("search.date", "date.transform")
for (k in dates) {
  datSource[,k] <- paste('="', as.character(datSource[,k]), '"', sep = '') 
}

datSource$search.date <- ifelse(datSource$date.transform == "=\"NA\"", 
                                as.character(datSource$search.date), 
                                datSource$date.transform)

### Data Cleaning (datSource)
# Remove internal use columns (not in specs), please ignore date.transform column
# The 2 default columns are: entry, qc
names(datSource)[1] <- "search"
names(datSource)
internalCols <- c("entry","QC") # adjust internal colnames as needed

datSource <- datSource %>%
  mutate(across(where(is.character), trimws),
         across(c(search, include, reason, e.copy, database), tolower)) %>%
  dplyr::select(!all_of(internalCols) & !date.transform)

names(datSource) # database should be the last column

### Subset needed comlumns (OutcomesDB)
OutcomesDB <- OutcomesDB %>%
  dplyr::select(url:publication.type,
                study,
                registry.study.id,
                n.study,
                arm,
                control.arm,
                treatment)

# DATA FORMAT CHECKS ----------------------------------------------------------------

# Check required columns 
checkReqSourceCols(datSource)
checkNumSourceCols(datSource)

# Use following for unallowed non-numeric entries. Please uncomment following 2 lines and change the colnames in x.
#x <- c("source.number", "publication.year")
#datSource[x] <- sapply(datSource[x], as.numeric)


# NAMED RANGES ------------------------------------------------------------------

# Run line-by-line
unique(datSource$search[!datSource$search %in% searchType])
unique(datSource$include[!datSource$include %in% yesNoUnclear])
unique(datSource$reason[!datSource$reason %in% reasonType])
unique(datSource$e.copy[!datSource$e.copy %in% yesNo])
unique(datSource$database[!datSource$database %in% yesNo])

# REFERENCES CHECKS -----------------------------------------------------------------

wDB.sNumber <- unique(OutcomesDB$source.number)
sDB.database.yes <- unique(datSource$source.number[datSource$database == "yes"])
sDB.database.no <- unique(datSource$source.number[datSource$database != "yes"])

# Check database = yes in source DB
wDB.no <- sDB.database.yes[!sDB.database.yes %in% wDB.sNumber]
checkSourceYes(wDB.no)

# Check database == no in source DB
wDB.yes <- sDB.database.no[sDB.database.no %in% wDB.sNumber]
datSource$database[datSource$source.number %in% wDB.yes] <- "yes"

# Check the rest of database == no (& include == yes) in Source DB
sDB.include.yes <- datSource %>%
  filter(include == "yes", database != "yes") %>%
  dplyr::select(source.number)
checkSourceNo(sDB.include.yes)

# Marked as pending if include == "yes"
datSource$database <- ifelse(datSource$source.number %in% sDB.include.yes, "pending", datSource$database)

# Marked as no
datSource$database <- ifelse(datSource$database == "", "no", datSource$database)

# CREATE DATASETS ----------------------------------------------------------------

### Create datasets for unique references, unique control drugs, and unique randomized drugs for studies ###

### Dataset - unique referenes that contains maximum arm for each study
datUniRef <- TrimData(OutcomesDB, source.number, study, arm)
datUniRef$arm <- gsub(".*\\+|[^0-9]", "", datUniRef$arm)
datUniRef <- datUniRef %>%
  group_by(source.number, study) %>%
  slice_max(order_by = arm, with_ties = FALSE, n = 1)

# Identify duplicated source.number for more than 1 study, if any
checkMultiStudy(datUniRef)
refs.multiple.study <- unique(datUniRef$source.number[duplicated(datUniRef$source.number)])

# Combine multiple studies information
datUniRef <- datUniRef %>%
  group_by(source.number) %>%
  mutate(across(c(study, registry.study.id, n.study, arm),
                function(x) paste(unique(x[!is.na(x) & x != ""]), collapse = " and "))) %>%
  filter(!duplicated(source.number))

### Dataset - unique control drugs dataset
datUniCtrl <- filter(OutcomesDB, !grepl("[A-z]", arm),            # exclude subarms
                     control.arm == "yes")                   # include control arms only
datUniCtrl <- TrimData(datUniCtrl, source.number, study, treatment)

datUniCtrl <- datUniCtrl %>%
  group_by(source.number) %>%
  filter(!duplicated(treatment)) %>%
  mutate(control.join = paste(treatment, collapse = ", ")) %>%
  filter(!duplicated(source.number)) %>%
  dplyr::select(source.number, control.join)


### Dataset - unique interventional drugs dataset
datUniTrt <- filter(OutcomesDB, !grepl("[A-z]", arm),             # exclude subarms
                    control.arm == "no")                     # include experimental arms only
datUniTrt <- TrimData(datUniTrt, source.number, study, treatment)

datUniTrt <- datUniTrt %>%
  group_by(source.number) %>%
  filter(!duplicated(treatment)) %>%
  mutate(treatment.join = paste(treatment, collapse = ", ")) %>%
  filter(!duplicated(source.number)) %>%
  dplyr::select(source.number, treatment.join)

# datUniRef and datUniTrt datasets should have the same x dimension \\\datUniCtrl could be missing due to single arm studies
dim(datUniRef)
dim(datUniCtrl)
dim(datUniTrt)
# can use the following code to find out which is missing in datUniTrt and look into augemented db
datUniRef$source.number[!datUniRef$source.number %in% datUniTrt$source.number]
# !!!Use the next chunk of codes if we have missing in datUniTrt!!!
 
# # Join treatment if the arms are subarms, strata only reference (have to check first in outcomes DB)
# tmp <- datUniRef$source.number[datUniRef$source.number %in% datUniTrt$source.number == FALSE]
# tmp.trt <- OutcomesDB %>%
#   filter(source.number %in% tmp) %>%
#   filter(!duplicated(treatment)) %>%
#   group_by(source.number) %>%
#   mutate(treatment.join = paste(treatment, collapse = ", ")) %>%
#   filter(!duplicated(source.number)) %>%
#   select(source.number, treatment.join)
# 
# datUniTrt <- rbind(datUniTrt, tmp.trt)

# FILL IN MISSING INFORMATION ---------------------------------------------------------------------------

# Merge datUniRef/datUniCtrl/datUniTrt
cols.join <- c("url",
               "authors",
               "publication.year",
               "title",
               "journal",
               "volume",
               "pages",
               "study",
               "registry.study.id",
               "n.study",
               "arm")

datUniRef.join <- datUniRef %>%
  dplyr::select(source.number, all_of(cols.join)) %>%
  rename(n.arm = arm) %>%
  rename_at(vars(-source.number), paste0, ".join") %>%
  list(datUniCtrl, datUniTrt) %>%
  reduce(left_join, by = "source.number")

# For oncology database single arm study (can also run it in general database)
datUniRef.join$control.join[is.na(datUniRef.join$control.join)] <- "no control"


### Check if it's in the augmented database, but not in the source database ###
miss.in.source <- wDB.sNumber[!wDB.sNumber %in% datSource$source.number]

# Add missing refs' source numbers to datUniRef.join for filling information later  
for (miss.info in miss.in.source) {
  datSource <- datSource %>%
    add_row(source.number = miss.info, database = "yes")
}


### All augmentated DB source.number should be in the source database now, proceed to fill in information ###
cols.join <- gsub(".join", "", names(datUniRef.join[, names(datUniRef.join) != c("source.number")]))
cols.join

datSource <- datSource %>%
  left_join(datUniRef.join, by = "source.number")

# Fill in info from augmented outcome DB
for(names in cols.join) {
  datSource[[names]] <- ReplaceInfo(names)
}

# URL HARMONIZATION ---------------------------------------------------------------------- (need improvements)

# Generate PubMed/clintrials URLs
datSource$url[datSource$search == "pubmed"] <- paste0("https://www.ncbi.nlm.nih.gov/pubmed/",
                                                      datSource$source.number[datSource$search == "pubmed"])
datSource$url[datSource$search == "clinicaltrials.gov"] <- paste0("https://clinicaltrials.gov/ct2/show/",
                                                                  datSource$registry.study.id[datSource$search == "clinicaltrials.gov"])
# Remove everything that comes after a comma in the URL
datSource$url <- sapply(strsplit(datSource$url, ","), "[", 1)

# CHECK INFORMATION FOR REFERENCES MORE THAN 1 STUDY -------------------------------------------------

check.multi.study <- datSource %>%
  filter(source.number %in% refs.multiple.study) %>%
  dplyr::select(source.number,
         study, 
         registry.study.id,
         treatment,
         control,
         n.study,
         n.arm)
view(check.multi.study)

# DATA FINAL CLEANING --------------------------------------------------------------------------------

#Unify study name format (remove quotation marks)
datSource$study <- gsub('=|"', "", datSource$study)

# Final cleaning and save to file
datSource <- dplyr::select(datSource, !ends_with("join"))

datSource$include <- ifelse(datSource$database == "yes", "yes", datSource$include)
datSource$include <- ifelse(datSource$database == "no", "no", datSource$include)

# Save as UTF-8 encoding file to avoid special characters in authors
write.csv(datSource, paste(dirResults,'Source database.csv',sep = '/'), na = "", row.names = FALSE, fileEncoding = "UTF-8")

# MESSAGES ---------------------------------------------------------------

# Find the position for any new information needed manually
#miss.in.source <- wDB.sNumber[!wDB.sNumber %in% datSource$source.number]
#wDB.sNumber[!wDB.sNumber %in% datSource$source.number]
ManualFill(miss.in.source)

# Find the position for missing URLs, if there is any
MissingURLs(datSource)

# Find if 
unique(datSource$database[!datSource$database %in% yesNo]) #should be empty

# END OF CODE -----------------------------------------------------------------------------
